v3.1/glfw/util.go

```diff
@@ -6,33 +6,41 @@ package glfw
 import "C"
 
 import (
-	"reflect"
-	"unsafe"
+	"image"
+	"image/draw"
 )
 
 func glfwbool(b C.int) bool {
-	if b == C.int(True) {
-		return true
-	}
-	return false
+	return b == C.int(True)
 }
 
 func bytes(origin []byte) (pointer *uint8, free func()) {
 	n := len(origin)
-
 	if n == 0 {
 		return nil, func() {}
 	}
 
-	data := C.malloc(C.size_t(n))
+	ptr := C.CBytes(origin)
+	return (*uint8)(ptr), func() { C.free(ptr) }
+}
+
+func imageToGLFW(img image.Image) (r C.GLFWimage, free func()) {
+	b := img.Bounds()
+
+	r.width = C.int(b.Dx())
+	r.height = C.int(b.Dy())
 
-	dataSlice := *(*[]byte)(unsafe.Pointer(&reflect.SliceHeader{
-		Data: uintptr(data),
-		Len:  n,
-		Cap:  n,
-	}))
+	if m, ok := img.(*image.NRGBA); ok && m.Stride == b.Dx()*4 {
+		end := m.PixOffset(m.Rect.Max.X, m.Rect.Max.Y)
+		pix, free := bytes(m.Pix[:end])
+		r.pixels = (*C.uchar)(pix)
+		return r, free
```

Comment:
Do we need to copy pixels? i.e., what about:

```go
return (*C.uchar)(unsafe.Pointer(&m.Pix[0])), runtime.KeepAlive(m)
```

?

Reply:
I did consider it, but I wasn't sure how the glfw image will be later used. If it's not going to be used in other scenarios later down the line then it should be fine. And as far as I understand, the functions using images seemed such that they aren't called every frame, so the allocation overhead probably won't matter.

Either way, I'll add the optimization.

Reply:
Done

Reply:
Yes.